File: app/src/main/java/com/litus_animae/refitted/data/ExerciseDao.java
package com.litus_animae.refitted.data;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import com.litus_animae.refitted.models.Exercise;
import com.litus_animae.refitted.models.ExerciseSet;
import com.litus_animae.refitted.models.SetRecord;

import java.util.Date;
import java.util.List;

@Dao
public interface ExerciseDao {
    @Query("select distinct step from exerciseset where day = :day and workout = :workout")
    LiveData<List<String>> getSteps(String day, String workout);

    @Query("select * from exerciseset where day = :day and workout = :workout and step = :step")
    ExerciseSet getExerciseSet(String day, String workout, String step);

    @Query("select * from exerciseset where day = :day and workout = :workout and step in (:steps) order by step")
    LiveData<List<ExerciseSet>> getExerciseSets(String day, String workout, String... steps);

    @Query("select * from exercise where exercise_name = :name and exercise_workout = :workout")
    Exercise getExercise(String name, String workout);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void storeExercise(Exercise exercise);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void storeExerciseSet(ExerciseSet exerciseSet);

    @Query("select * from setrecord where completed > :minDate and exercise = :targetExercise")
    List<SetRecord> getSetRecords(Date minDate, String targetExercise);

    @Query("select * from setrecord where exercise = :targetExercise order by completed desc")
    SetRecord getLatestSetRecord(String targetExercise);

    @Query("select * from setrecord where exercise = :targetExercise order by completed desc")
    List<SetRecord> getAllSetRecord(String targetExercise);

    @Insert
    void storeExerciseRecord(SetRecord exerciseRecord);
}

File: app/src/main/java/com/litus_animae/refitted/data/ExerciseDao.kt
package com.litus_animae.refitted.data

import androidx.lifecycle.LiveData
import androidx.paging.DataSource
import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.litus_animae.refitted.models.Exercise
import com.litus_animae.refitted.models.ExerciseSet
import com.litus_animae.refitted.models.SetRecord
import java.util.*

@Dao
interface ExerciseDao {
    @Query("select distinct step from exerciseset where day = :day and workout = :workout")
    fun getSteps(day: String, workout: String): LiveData<List<String>>

    @Query("select * from exerciseset where day = :day and workout = :workout and step = :step")
    fun getExerciseSet(day: String, workout: String, step: String): ExerciseSet?

    @Query("select * from exerciseset where day = :day and workout = :workout and step in (:steps) order by step")
    fun getExerciseSets(day: String, workout: String, vararg steps: String?): LiveData<List<ExerciseSet>>

    @Query("select * from exercise where exercise_name = :name and exercise_workout = :workout")
    fun getExercise(name: String, workout: String): LiveData<Exercise>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun storeExercise(exercise: Exercise)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun storeExerciseSet(exerciseSet: ExerciseSet)

    @Query("select * from setrecord where completed > :minDate and exercise = :targetExercise")
    fun getSetRecords(minDate: Date, targetExercise: String): LiveData<List<SetRecord>>

    @Query("select * from setrecord where exercise = :targetExercise order by completed desc")
    fun getLatestSetRecord(targetExercise: String): LiveData<SetRecord?>

    @Query("select * from setrecord where exercise = :targetExercise order by completed desc")
    fun getAllSetRecord(targetExercise: String): DataSource.Factory<Int, SetRecord?>

    @Insert
    fun storeExerciseRecord(exerciseRecord: SetRecord)
}